### ISSUE TITLE
[XLA:CPU] Wrong results / out-of-bounds read in f32 convolution with negative low padding

### ISSUE THREAD

**Issue by EylonKrause** [openxla/xla#48713]
A `conv_general_dilated` on f32 whose spatial dimension has **negative low padding** and a
positive high padding of at most `|low|` reads outside the input buffer: the output positions that
should be zero padding contain data from the next row, and past the end of the buffer, uninitialised
memory. The identical convolution in `int32` (which goes through the elemental emitter rather than
Eigen) is correct, as is the `>16`-channel case that takes the YNNPACK path.

### Reproducer

```python
import os
os.environ["JAX_PLATFORMS"] = "cpu"
import numpy as np, jax, jax.numpy as jnp
from jax import lax

H, W, C = 15, 6, 8
x = (100 * np.arange(H)[:, None, None] + 10 * np.arange(W)[None, :, None] + np.arange(C)[None, None, :])
x = x.reshape(1, H, W, C)                                  # x[0,h,w,c] = 100h + 10w + c  (NHWC)
k = np.eye(C).reshape(1, 1, C, C)                          # 1x1 identity kernel (HWIO): out[...,c] = padded_x[...,c]
dn = lax.ConvDimensionNumbers((0, 3, 1, 2), (3, 2, 0, 1), (0, 3, 1, 2))
pads = ((0, 0), (-5, 1))                                   # W: crop 5 -> [x5], then pad one zero -> [x5, 0]

for dtype in (jnp.float32, jnp.int32):
    f = lambda a, b: lax.conv_general_dilated(a, b, (1, 1), pads, dimension_numbers=dn)
    xa, ka = jnp.asarray(x, dtype), jnp.asarray(k, dtype)
    expected = np.stack([x[0, :, 5, :], np.zeros((H, C))], axis=1)[None]      # (1, 15, 2, 8)
    eager = np.asarray(f(xa, ka)); jitted = np.asarray(jax.jit(f)(xa, ka))
    print(dtype.__name__, "row0 expected:", expected[0, 0].tolist())
    print(dtype.__name__, "row0 eager   :", eager[0, 0].tolist(), "WRONG" if not np.array_equal(eager, expected) else "ok")
    print(dtype.__name__, "row0 jit     :", jitted[0, 0].tolist(), "WRONG" if not np.array_equal(jitted, expected) else "ok")
    print(dtype.__name__, "last row jit  :", jitted[0, -1, 1].tolist(), "(expected zeros)")
```

Output on my machine:

```
float32 row0 expected: [[50.0, ..., 57.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]
float32 row0 eager   : [[50.0, ..., 57.0], [100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0]] WRONG
float32 row0 jit     : [[50.0, ..., 57.0], [100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0]] WRONG
float32 last row jit : [0.0, 0.0, 0.0, 0.0, -4891756.0, 0.0, 0.0, 0.0]   (expected zeros)
int32   row0 eager   : [[50, ..., 57], [0, 0, 0, 0, 0, 0, 0, 0]] ok
int32   row0 jit     : [[50, ..., 57], [0, 0, 0, 0, 0, 0, 0, 0]] ok
```

### Analysis

With a 1x1 identity kernel the output is just the padded input, so the padded `W` axis must be
`[x[:, 5, :], 0]`. Instead the second position holds `x[0, h+1, 0, :]`, and in the last row it holds
whatever follows the buffer (`-4891756.0` here; other runs gave `-3.7e27`, `4.9e36`).

The suspect is the `hasPadding()` helper used by the Eigen image-patch path
(`xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h`), which only considers padding
*positive*, so a negative before-padding leaves the "no padding" fast path selected and the patch
extraction indexes the input directly.

Affected padding pairs I checked on the `W` axis: `(-6,1) (-6,2) (-5,1) (-5,2) (-4,3) (-3,1) (-7,3)
(-8,3)`, and on `H`: `(-15,1) (-15,2) (-14,2)`. A positive low padding with a negative high padding
(`(1,-6)`, `(2,-7)`) is fine. Negative convolution padding is a documented feature and JAX emits it
in convolution VJPs, so this is reachable from ordinary gradient code.

### Environment

- `jax` 0.11.1, `jaxlib` 0.11.1 (pip), CPU backend, `JAX_PLATFORMS=cpu`
- Linux x86-64 (AVX-512), Python 3.12

### Notes

Found by a differential fuzzer comparing XLA:CPU against a NumPy oracle written from the operation semantics (validated by hand and against `jax._src.lax_reference`), then minimised. I am filing this as an issue rather than a pull request, per the repository's contribution guidelines: this report was prepared with the help of an AI coding assistant (Claude), and I reproduced every result above myself on the environment listed.

**Comment by EylonKrause** [comment]
Follow-up: the same path also **segfaults** when the crop removes the whole dimension, so this is a memory-safety issue and not only a wrong-value one.

```python
import os
os.environ["JAX_PLATFORMS"] = "cpu"
import numpy as np, jax, jax.numpy as jnp
from jax import lax

C, H, W = 17, 15, 6
x = jnp.asarray(np.arange(H * W * C).reshape(1, H, W, C), jnp.float32)
k = jnp.asarray(np.eye(C).reshape(1, 1, C, C), jnp.float32)          # 1x1 identity
dn = lax.ConvDimensionNumbers((0, 3, 1, 2), (3, 2, 0, 1), (0, 3, 1, 2))
f = lambda u, v: lax.conv_general_dilated(u, v, (1, 1), [(-15, 1), (0, 0)],
                                          dimension_numbers=dn)
print(np.asarray(jax.jit(f)(x, k)).shape)   # expected (1, 1, 6, 17), all zeros
```

The padding crops all 15 rows of `H` and then pads one, so the output is a single row that must be entirely zero padding. Instead the process dies with `SIGSEGV` (exit 139).

Splitting the steps shows it is the **generated kernel**, not the compiler: `lower()` and `compile()` both succeed and the crash happens on the first call.

```
lowering...
lowered ok
compiled ok
running...
Segmentation fault (core dumped)
```

Channel count is the switch, with the boundary between 16 and 17:

| channels `C` | result |
|---|---|
| 8, 12, 16 | runs, returns the wrong values reported above |
| 17, 20, 24, 32, 64 | **SIGSEGV** |

`((-15, 2), (0, 0))` (two output rows) runs, and `((-14, 1), (0, 0))` runs, so it is specifically the crop-everything-then-pad case.

This looks like the same root cause as the report above: with the before-padding ignored, the patch extraction starts `15` rows past the end of the image, which is a wrong-but-in-range read for a small channel count and off the end of the buffer for the wider vectorised path.
